isValueType and isNonValueType

paulp opened the pull request:
This should supersede #1384.  Review by @xeno-by, @retronym.

xeno-by committed:
showRaw no longer crashes on NoSymbol

xeno-by committed:
SI-6417 correctly reifies non-value types

If we're reifying non-value types (e.g. MethodTypes), we can't use them
as type arguments for TypeTag/WeakTypeTag factory methods, otherwise
the macro expansion won't typecheck:
http://groups.google.com/group/scala-internals/browse_thread/thread/2d7bb85bfcdb2e2

This situation is impossible if one uses only reify and type tags, but
c.reifyTree and c.reifyType exposes in the macro API let anyone feed
anything into the reifier.

Therefore I now check the tpe that is about to be used in TypeApply
wrapping TypeTag/WeakTypeTag factory methods and replace it with AnyTpe
if it doesn't fit.

paulp committed:
Implementations of isValueType and isNonValueType.

Restrictions regarding how non-value types can be used have
generally not been enforced explicitly, depending instead on
the fact that the compiler wouldn't attempt to use them in
strange ways like offering a method type as a type argument.
Since users can now create most types from scratch, it has
become important to enforce the restrictions in a more
direct fashion.

This was a lot harder than it probably should have been
because there are so many types which go unmentioned by the
specification.  Hopefully a useful exercise in any case.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/640/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1348/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/640/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1348/

paulp committed:
Fix for failing test.

There's some very sketchy behavior visible - I'm printing a
method signature and getting this:

  [B <: <?>, That <: <?>](f: <?>)(implicit cbf: <?>)That

But there's no exposed way to force the info.  Am I
supposed to call isSealed or something?

paulp commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/641/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1349/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1349/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/641/

xeno-by commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -5732,6 +5744,100 @@ trait Types extends api.Types { self: SymbolTable =>
+  )
+
+  private def isHKTypeRef(tp: Type) = tp match {
+    case TypeRef(_, sym, Nil) => tp.isHigherKinded
```
Can type refs with non-empty args return `true` from `tp.isHigherKinded`?

xeno-by commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -5732,6 +5744,100 @@ trait Types extends api.Types { self: SymbolTable =>
+    case tp if isAlwaysValueType(tp)           => true
+    case tp if isAlwaysNonValueType(tp)        => false
+    case AnnotatedType(_, underlying, _)       => isValueElseNonValue(underlying)
+    case SingleType(_, sym)                    => sym.isValue           // excludes packages and statics
```
Could you please give an example of a static this check excludes?

xeno-by commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -5732,6 +5744,100 @@ trait Types extends api.Types { self: SymbolTable =>
+    case TypeRef(_, _, _) if tp.isHigherKinded => false                 // excludes type constructors
+    case ThisType(sym)                         => !sym.isPackageClass   // excludes packages
+    case TypeRef(_, sym, _)                    => !sym.isPackageClass   // excludes packages
+    case PolyType(_, _)                        => true                  // poly-methods excluded earlier
```
What's an example of a `PolyType` which is a value type?

gkossakowski commented:
Cannot be merged. Also, needs review.

xeno-by commented:
forked in #1415, rebased against fresh trunk, fixed the failing test

xeno-by commented:
will try to catch Martin to get a review
